ctrd/client.go

```diff
@@ -166,3 +190,143 @@ func (c *Client) Version(ctx context.Context) (containerd.Version, error) {
 
 	return cli.client.Version(ctx)
 }
+
+func (c *Client) runContainerdDaemon(homeDir string, copts clientOpts) error {
+	if homeDir == "" {
+		return fmt.Errorf("ctrd: containerd home dir should not be empty")
+	}
+
+	containerdPath, err := exec.LookPath(copts.containerdBinary)
+	if err != nil {
+		return fmt.Errorf("failed to find containerd binary %s: %v", copts.containerdBinary, err)
+	}
+
+	stateDir := path.Join(homeDir, "containerd/state")
+	if _, err := os.Stat(stateDir); err != nil && os.IsNotExist(err) {
+		if err := os.MkdirAll(stateDir, 0666); err != nil {
+			return fmt.Errorf("failed to mkdir %s: %v", stateDir, err)
+		}
+	}
+
+	pidFileName := path.Join(stateDir, containerdPidFileName)
+	f, err := os.OpenFile(pidFileName, os.O_RDWR|os.O_CREATE, 0600)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	buf := make([]byte, 8)
+	num, err := f.Read(buf)
+	if err != nil && err != io.EOF {
+		return err
+	}
+
+	if num > 0 {
+		pid, err := strconv.ParseUint(string(buf[:num]), 10, 64)
+		if err != nil {
+			return err
+		}
+		if utils.IsProcessAlive(int(pid)) {
+			logrus.Infof("ctrd: previous instance of containerd still alive (%d)", pid)
+			c.daemonPid = int(pid)
+			return nil
+		}
+	}
+
+	// empty container pid file
+	_, err = f.Seek(0, os.SEEK_SET)
+	if err != nil {
+		return err
+	}
+
+	if err := f.Truncate(0); err != nil {
+		return err
+	}
+
+	// if socket file exists, delete it.
+	if _, err := os.Stat(c.rpcAddr); err == nil {
+		os.RemoveAll(c.rpcAddr)
+	}
+
+	cmd, err := c.newContainerdCmd(containerdPath)
+	if err != nil {
+		return err
+	}
+
+	if err := utils.SetOOMScore(cmd.Process.Pid, c.oomScoreAdjust); err != nil {
+		utils.KillProcess(cmd.Process.Pid)
+		return err
+	}
+
+	if _, err := f.WriteString(fmt.Sprintf("%d", cmd.Process.Pid)); err != nil {
+		utils.KillProcess(cmd.Process.Pid)
+		return err
+	}
+
+	go cmd.Wait()
+
+	c.daemonPid = cmd.Process.Pid
+	return nil
+}
+
+func (c *Client) newContainerdCmd(containerdPath string) (*exec.Cmd, error) {
+	// Start a new containerd instance
+	args := []string{
+		"-a", c.rpcAddr,
+		"--root", path.Join(c.homeDir, "containerd/root"),
```

Comment:
Sorry. I don't make my suggestion clear in the last comment. 

The containerd has own the configuration. The link is [here](https://github.com/containerd/containerd/blob/master/server/config.go#L43). Maybe we can marshal the settings in the file so that we can use the file to check the information about pouchd.

Reply:
actually, the containerd core `Opts` have  been in the poucd configuration file,  and i think we no need keep much containerd configuration opts in pouchd code, if you want to add more configurations , use the containerd configuration file outside, DYA?

Reply:
LGTM